workflow/util/util.go

```diff
@@ -1049,16 +1058,18 @@ func GetTemplateFromNode(node wfv1.NodeStatus) string {
 
 func getNodeIDsToReset(restartSuccessful bool, nodeFieldSelector string, nodes wfv1.Nodes) (map[string]bool, error) {
 	nodeIDsToReset := make(map[string]bool)
-	if !restartSuccessful || len(nodeFieldSelector) == 0 {
+	if len(nodeFieldSelector) == 0 {
 		return nodeIDsToReset, nil
 	}
-
 	selector, err := fields.ParseSelector(nodeFieldSelector)
 	if err != nil {
 		return nil, err
 	} else {
 		for _, node := range nodes {
 			if SelectorMatchesNode(selector, node) {
+				if !restartSuccessful && node.Phase == wfv1.NodeSucceeded {
+					return nil, errors.Errorf(errors.CodeBadRequest, "cannot restart successful node %s , plz add --restart-successful", node.Name)
```

Comment:
this is a helper function, it shouldn't error out when you pass 1 of 2 options (it's a boolean) of an argument to it.

this should just `continue` if that's the case, which properly handles both conditions
```suggestion
					continue
```

also this is a formal error message, it should not use language like "plz"
and it's an API error message, while `--restart-successful` is specific to the CLI

Reply:
the `continue` is a correctness fix, but this would be better consolidated if it actually used the selector to filter out this condition instead

Reply:
Above comment followed up on [below](https://github.com/argoproj/argo-workflows/pull/12553#discussion_r1483696538)